### PR TITLE
move probe, session logic, flash downloader to probe-rs-cli-util

### PR DESCRIPTION
**Pull request by tmplt**
See https://github.com/probe-rs/probe-rs/pull/723.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -411,74 +210,45 @@ fn main_try() -> Result<(), CargoFlashError> {
...
+        log::info!("Protocol speed {} kHz", protocol_speed);
+    }
 
-    // Create a new session.
-    // If we wanto attach under reset, we do this with a special function call.
-    // In this case we assume the target to be known.
-    // If we do an attach without a hard reset, we also try to automatically detect the chip at hand to improve the userexperience.
-    let mut session = if opt.connect_under_reset {
+    // Create a new session
+    let mut session = if opt.probe_options.connect_under_reset {
```
Comment (tmplt) [inline src/main.rs:239]: This logic should be moved to `probe-rs-cli-util` but that would require us to store the `Probe` in `common_options` such that the API is somewhat ergonomic. It is not possible to store non-structopt fields in a `#[structopt] struct { .. }`; i.e. a `Option<Probe>`.

A `lazy_static` maybe?

File: src/main.rs [hunk 2]
```diff
@@ -491,59 +261,32 @@ fn main_try() -> Result<(), CargoFlashError> {
...
-    let loader = match loader {
-        Some(loader) => loader,
-        None => build_flashloader(session.target(), path)?,
-    };
+) -> Result<(), OperationError> {
+    // let loader = match loader {
+    //     Some(loader) => loader,
+    //     None => build_flashloader(session.target(), path)?,
+    // };
+    let loader = loader.unwrap();
```
Comment (tmplt) [inline src/main.rs:289]: To be handled correctly.

**Comment by tmplt** [comment]
`ARGUMENTS_TO_REMOVE` should also be moved to `probe-rs-cli-util`.

**Comment by tmplt** [comment]
CC @Tiwalun

File: src/main.rs (hunk 1 above)
Comment (tmplt) [reply src/main.rs]: Done via `static mut` for now.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -25,7 +25,7 @@ jobs:
         uses: actions/checkout@v2.3.4
 
       - name: Install libusb
-        run: sudo apt install -y libusb-1.0-0-dev libftdi1-dev
+        run: sudo apt install -y libusb-1.0-0-dev libftdi1-dev pkg-config
```
Comment (Tiwalun) [inline .github/workflows/ci.yml:28]: The issue is probably missing `libudev`, `pkg-config` should already be installed.
The following should work:

```yaml
- name: Install libusb, libudev and libftdi (linux)
        run: |
          sudo apt update
          sudo apt install -y libusb-1.0-0-dev libftdi1-dev libudev-dev
        # Only install on Ubuntu
        if: (matrix.os == 'ubuntu-latest') || (matrix.os == 'ubuntu-20.04')
```

(Taken from https://github.com/probe-rs/probe-rs/blob/master/.github/workflows/ci.yml)

**Comment by tmplt** [comment]
Converting to draft. I don't think arguments to cargo passes structopt correctly.

**Comment by tmplt** [comment]
> I don't think arguments to cargo passes structopt correctly.

I just fixed this in `cargo-rtic-scope`. In short, arguments that should forward to cargo must passed as the last options on the CLI. See https://github.com/clap-rs/clap/issues/2354.

**Comment by tmplt** [comment]
Do we want to continue with this so as not to block any future pull requests? We could add a diagnostic entry when clap fails and tell the user to move the cargo arguments to the end.

**Comment by tmplt** [comment]
Forwaring tail-end arguments to cargo has had some [unexpected side-effects in cargo-rtic-scope](https://github.com/rtic-scope/cargo-rtic-scope/issues/30), presuming that I have applied clap settings correctly.

A possible alternative is to use a sentinel `--`; any arguments after it can be passed to cargo. But perhaps it's better to revert https://github.com/probe-rs/probe-rs/commit/5075192ab2ad8f056c4be5bcc040ed7caed09d17 until there is proper support for this in clap?

**Comment by Tiwalun** [comment]
I think we could add the struct with the common cargo options, and maybe add the sentinel `--` as an escape hatch, so if someone wants to use an unsupported argument, it's still possible? Not sure how easy this is to implement, however.

**Review by Yatekii — APPROVED** [review]
Thanks!

bors r+

**Comment by tmplt** [comment]
CC @Yatekii. Master is broken until this merges.

**Pull request merged**